src/main/java/org/prebid/server/bidder/gamoshi/GamoshiBidder.java

```diff
@@ -0,0 +1,174 @@
+            final Banner modifiedBanner = banner.toBuilder()
+                    .h(firstFormat.getH())
+                    .w(firstFormat.getW())
+                    .build();
+            return imp.toBuilder().banner(modifiedBanner).build();
+        }
+        return imp;
+    }
+
+    private static ExtImpGamoshi parseAndValidateImpExt(Imp imp) {
+        final ExtImpGamoshi extImpGamoshi;
+        try {
+            extImpGamoshi = Json.mapper.convertValue(imp.getExt().get("bidder"), ExtImpGamoshi.class);
+        } catch (IllegalArgumentException e) {
+            throw new PreBidException(e.getMessage(), e);
+        }
+
+        if (StringUtils.isBlank(extImpGamoshi.getSupplyPartnerId())) {
+            throw new PreBidException("supplyPartnerId is empty");
+        }
+
+        return extImpGamoshi;
+    }
+
+    private static MultiMap resolveHeaders(Device device) {
+        final MultiMap headers = BidderUtil.headers()
+                .add("x-openrtb-version", "2.4");
+        if (device != null) {
+            HttpUtil.addHeaderIfValueIsNotEmpty(headers, "User-Agent", device.getUa());
+            HttpUtil.addHeaderIfValueIsNotEmpty(headers, "X-Forwarded-For", device.getIp());
+            HttpUtil.addHeaderIfValueIsNotEmpty(headers, "Accept-Language", device.getLanguage());
+            HttpUtil.addHeaderIfValueIsNotEmpty(headers, "DNT", device.getDnt().toString());
+        }
+        return headers;
+    }
+
+    @Override
+    public Result<List<BidderBid>> makeBids(HttpCall<BidRequest> httpCall, BidRequest bidRequest) {
+        try {
+            final BidResponse bidResponse = Json.decodeValue(httpCall.getResponse().getBody(), BidResponse.class);
+            return Result.of(extractBids(httpCall.getRequest().getPayload(), bidResponse), Collections.emptyList());
+        } catch (DecodeException | PreBidException e) {
+            return Result.emptyWithError(BidderError.badServerResponse(e.getMessage()));
+        }
+    }
+
+    private static List<BidderBid> extractBids(BidRequest bidRequest, BidResponse bidResponse) {
+        if (bidResponse == null || bidResponse.getSeatbid() == null) {
+            return Collections.emptyList();
+        }
+        return bidsFromResponse(bidRequest, bidResponse);
+    }
+
+    private static List<BidderBid> bidsFromResponse(BidRequest bidRequest, BidResponse bidResponse) {
+        return bidResponse.getSeatbid().stream()
+                .filter(Objects::nonNull)
+                .map(SeatBid::getBid)
+                .filter(Objects::nonNull)
+                .flatMap(Collection::stream)
+                .map(bid -> BidderBid.of(bid, getBidType(bid.getImpid(), bidRequest.getImp()), DEFAULT_BID_CURRENCY))
+                .collect(Collectors.toList());
+    }
+
+    private static BidType getBidType(String impId, List<Imp> imps) {
+        for (Imp imp : imps) {
+            if (imp.getId().equals(impId) && imp.getVideo() != null) {
```

Comment:
We can do refactor like:
```
if (impId.equals(imp.getId())) {
  return imp.getVideo() != null ? BidType.video : BidType.banner;
}
```
b/c we don't need to go to end of list if we have already found imp by ID.